docs/hugo/content/architecture-decision-records/ADR-2022-01-Reconciler-Extensions.md

```diff
@@ -0,0 +1,123 @@
+
+### Example
+
+To allow customization of error handling for Azure Redis, we'll introduce an extension point, which would look something like this:
+
+We create the required extension interface in `genruntime`:
+
+``` go
+type ErrorClassifierExtension interface {
+    // Classify the provided error, returning details about the classification.
+    // cloudError is the error returned from ARM.
+    // apiVersion is the ARM API version used for the request
+    // next is the function to call for standard classification behaviour
+    ClassifyError(
+        cloudError *genericarmclient.CloudError, 
+        apiVersion string, 
+        next func(*genericarmclient.CloudError) (*CloudErrorDetails, error)) (*CloudErrorDetails, error)
+}
+```
+
+In the `cache/extensions` package, a host type for the extension point will be generated in `redis_extensions_gen.go`:
+
+```
+type RedisExtensions struct{}
+```
+
+**TODO**: Should this have any members?
+
+Manual implementation of the extension point will be in `redist_extensions.go` (a different file so that it's not obliterated next time we re-run the generator):
+
+``` go
+var _ ErrorClassifierExtension = &RedisExtensions{}
+
+func (e *RedisExtensions) ClassifyError(
+    cloudError *genericarmclient.CloudError, 
+    next func(*genericarmclient.CloudError) *CloudErrorDetails) (*CloudErrorDetails, error)
+
+    err, result := next(cloudError)
+    if err != nil {
+        return nil, err
+    }
+
+    if inner := cloudError.InnerError; inner != nil {
+        code := to.String(inner.Code)
+        if code == "Conflict" {
+            // For Redis, Conflict can be retried as it can occur because something doesn't yet exist
+            result.Classification = CloudErrorRetryable
+        }
+    }
+
+    return result, nil
+}
+```
+
+Passing in the default behavour as `next` allows extensions to augmenet and/or supplant the behaviour with a great deal of flexibility.
+
+In the generic reconciler, we make use of the extension:
+
+``` go
+func (r *AzureDeploymentReconciler) makeReadyConditionFromError(
+    cloudError *genericarmclient.CloudError) conditions.Condition {
+    // ... existing code elided ...
+
+    ext := r.FindErrorClassifierExtension()
+    errorDetails, err := ext.ClassifyError(cloudError, apiVersion, r.ClassifyCloudError)
+    if err != nil {
+        // handle error
+    }
+
+    // ... existing code elided ...
+}
+```
+
+The helper method `FindErrorClassifierExtension()` does the lookup for the extension, and returns a null implementation if not found, so the consuming code doesn't need to worry about null checking.
+
+**TODO**: Should we actually return a wrapper around the extension that provides logging and other useful features?
```

Comment:
Might need to pass some "standard stuff" to the extensions, for example logging, Kubernetes client, Azure client, etc.

Not sure if that's a "wrapper" per-se, but some sort of common `ExtensionEssentials` datatype holding `log`, `armClient`, `kubeClient`, etc?

Reply:
I was more thinking that the implementation of `ErrorClassifierExtension` returned would be one that does logging itself, with the `RedisExtensions` instance wrapped inside it; that way individual extensions don't need to worry about logging.

Would it ever be sensible for clients to be passed to an extension? They're supposed to tweak the reconciler's behaviour, not provide an opportunity for doing something completely new.

Reply:
If we wanna do secrets via the extension mechanism they'll need both an Azure client and a k8s client, because they will have to:

1. Call the Azure `GetSecrets API`
2. Write the k8s (or KeyVault or whatever) secret.

We could do the secrets stuff in its own special extension though, and deny others access to the clients. I'm not sure if there are other situations where we'll want access to one or the other clients in extensions - do we have a catalog of the customizations we know we'd like to perform? We could look through it and see if they're all simple or if some are going to require calls to Azure.

Reply:
Each extension point has a custom set of parameters, so we can pass those clients through for that extension point without offering them anywhere else. I've clarified the spec to make this clearer.

Reply:
Makes sense. I've clarified in the spec that each extension is custom designed.